### PR TITLE
[Sparse] Add reduction operators

### PR DESCRIPTION
**Pull request by czkkkkkk**
## Description
Migrate sparse reduction operators to DGL repo.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: python/dgl/mock_sparse/reduce.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+    does not have any non-zero value, the result will be 0.
+
+    Parameters
+    ----------
+    A : SparseMatrix
+        Sparse matrix
+    dim : int, optional
+        The dimension to reduce.
+    rtype: str
+        Reduction type, one of ['sum', 'smin', 'smax', 'mean']
```
Comment (mufeili) [inline python/dgl/mock_sparse/reduce.py:24]: Perhaps explain a bit about `smin` and `smax`

File: python/dgl/mock_sparse/reduce.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+    Tensor
+        Reduced tensor
+
+    Examples
+    ----------
+
+    >>> row = torch.tensor([0, 1, 1])
+    >>> col = torch.tensor([0, 0, 2])
+    >>> val = torch.tensor([1, 1, 2])
+    >>> A = SparseMatrix(row, col, val, shape=(4, 3))
```
Comment (mufeili) [inline python/dgl/mock_sparse/reduce.py:37]: Use `create_from_coo` instead

File: python/dgl/mock_sparse/reduce.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+    tensor(4)
+    >>> print(A.sum(0))
+    tensor([2, 0, 2])
+    >>> print(A.sum(1))
+    tensor([1, 3, 0, 0])
+
+    >>> row = torch.tensor([0, 1, 1])
+    >>> col = torch.tensor([0, 0, 2])
+    >>> val = torch.tensor([[1, 2], [2, 1], [2, 2]])
+    >>> A = SparseMatrix(row, col, val, shape=(4, 3))
```
Comment (mufeili) [inline python/dgl/mock_sparse/reduce.py:129]: Ditto

File: python/dgl/mock_sparse/reduce.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+    Tensor
+        Reduced tensor
+
+    Examples
+    ----------
+
+    >>> row = torch.tensor([0, 1, 1])
+    >>> col = torch.tensor([0, 0, 2])
+    >>> val = torch.tensor([1, 1, 2])
+    >>> A = SparseMatrix(row, col, val, shape=(4, 3))
```
Comment (mufeili) [inline python/dgl/mock_sparse/reduce.py:164]: Ditto

File: python/dgl/mock_sparse/reduce.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+    >>> row = torch.tensor([0, 1, 1])
+    >>> col = torch.tensor([0, 0, 2])
+    >>> val = torch.tensor([[1, 2], [2, 1], [2, 2]])
+    >>> A = SparseMatrix(row, col, val, shape=(4, 3))
+    >>> print(A.max())
+    tensor([2, 2])
+    >>> print(A.max(0))
+    tensor([[2, 2], [0, 0], [2, 2]])
+    >>> print(A.max(1))
```
Comment (mufeili) [inline python/dgl/mock_sparse/reduce.py:180]: I guess all these `max` should be `smax`.

File: python/dgl/mock_sparse/reduce.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+
+    >>> row = torch.tensor([0, 1, 1])
+    >>> col = torch.tensor([0, 0, 2])
+    >>> val = torch.tensor([[1, 2], [2, 1], [2, 2]])
+    >>> A = SparseMatrix(row, col, val, shape=(4, 3))
+    >>> print(A.min())
+    tensor([1, 1])
+    >>> print(A.min(0))
+    tensor([[1, 1], [0, 0], [2, 2]])
+    >>> print(A.min(1))
```
Comment (mufeili) [inline python/dgl/mock_sparse/reduce.py:228]: I guess all these `min` should be `smin`.

File: python/dgl/mock_sparse/reduce.py (hunk 1 above)
Comment (mufeili) [inline python/dgl/mock_sparse/reduce.py:24]: Why we have `smin` and `smax`, not `ssum` and `smean`?

File: tests/pytorch/mock_sparse/test_reduce.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+@pytest.mark.parametrize("dense_dim", [None, 2])
+@pytest.mark.parametrize("row", [[0, 0, 1, 2], (0, 1, 2, 4)])
+@pytest.mark.parametrize("col", [(0, 1, 2, 2), (1, 3, 3, 4)])
+@pytest.mark.parametrize("mat_shape", [(3, 5), (5, 3)])
+@pytest.mark.parametrize("reduce_type", ['sum', 'smax', 'smin', 'mean'])
+@pytest.mark.parametrize("dim", [None, 0, 1])
+def test_reduce(dense_dim, row, col, mat_shape, reduce_type, dim):
+    # Skip invalid matrices
+    if max(row) >= mat_shape[0] or max(col) >= mat_shape[1]:
```
Comment (mufeili) [inline tests/pytorch/mock_sparse/test_reduce.py:16]: Perhaps it's better to replace `mat_shape` with `extra_shape` or `shape_offset` for expanding the shape dictated with `row` and `col`.

File: tests/pytorch/mock_sparse/test_reduce.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+            else:
+                if reduce_type == 'mean':
+                    reduced[i] /= count[i]
+        return numpy.stack(reduced, axis=0)
+    expected = calc_expected(row, col, val, mat_shape, reduce_type, dim)
+
+    assert torch.allclose(reduced, torch.tensor(expected).float())
+        
+if __name__ == "__main__":
+    test_reduce()
```
Comment (mufeili) [inline tests/pytorch/mock_sparse/test_reduce.py:75]: No need for the last two lines

**Comment by jermainewang** [comment]
@dgl-bot

File: python/dgl/mock_sparse/reduce.py (hunk 1 above)
Comment (czkkkkkk) [reply python/dgl/mock_sparse/reduce.py]: Done.

File: python/dgl/mock_sparse/reduce.py (hunk 2 above)
Comment (czkkkkkk) [reply python/dgl/mock_sparse/reduce.py]: Done

File: python/dgl/mock_sparse/reduce.py [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+    Tensor
+        Reduced tensor
+
+    Examples
+    ----------
+
+    >>> row = torch.tensor([0, 1, 1])
+    >>> col = torch.tensor([0, 0, 2])
+    >>> val = torch.tensor([1, 1, 2])
+    >>> A = SparseMatrix(row, col, val, shape=(4, 3))
```
[the remark this answers is not in the document]
Comment (czkkkkkk) [reply python/dgl/mock_sparse/reduce.py]: Done

File: python/dgl/mock_sparse/reduce.py (hunk 5 above)
Comment (czkkkkkk) [reply python/dgl/mock_sparse/reduce.py]: Fixed.

File: python/dgl/mock_sparse/reduce.py (hunk 1 above)
Comment (czkkkkkk) [reply python/dgl/mock_sparse/reduce.py]: Right. I added `smean` but not `ssum` because `0` does not affect the result.

File: tests/pytorch/mock_sparse/test_reduce.py (hunk 8 above)
Comment (czkkkkkk) [reply tests/pytorch/mock_sparse/test_reduce.py]: Done

File: tests/pytorch/mock_sparse/test_reduce.py (hunk 7 above)
Comment (czkkkkkk) [reply tests/pytorch/mock_sparse/test_reduce.py:16]: I didn't get your point. Basically I think mat_shape cannot be inferred from `row` and `col`.

Comment (mufeili) [reply tests/pytorch/mock_sparse/test_reduce.py:16]: Basically, we have

```
base_shape = (max(row) + 1, max(col) + 1)
target_shape = base_shape + offset_shape
```

Comment (czkkkkkk) [reply tests/pytorch/mock_sparse/test_reduce.py:16]: I see. I will change them accordingly.

File: python/dgl/mock_sparse/__init__.py [hunk 10]
```diff
@@ -2,3 +2,4 @@
 from .diag_matrix import *
 from .sp_matrix import *
 from .elementwise_op_sp import *
+from .reduce import * # pylint: disable=W0622
```
Comment (jermainewang) [inline python/dgl/mock_sparse/__init__.py:5]: Rename it to `reduction.py` or `reduce_op.py`. Try to avoid turn off lint as much as possible.

File: python/dgl/mock_sparse/reduce.py [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+    >>> A = create_from_coo(row, col, val, shape=(4, 3))
+    >>> print(A.sum())
+    tensor([5, 5])
+    >>> print(A.sum(0))
+    tensor([[3, 3], [0, 0], [2, 2]])
+    """
+    return A.reduce(dim, rtype="sum")
+
+
+def smax(A: SparseMatrix, dim=None):
```
Comment (jermainewang) [inline python/dgl/mock_sparse/reduce.py:144]: miss type annotation for `dim`

File: python/dgl/mock_sparse/reduce.py [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+    """Returns the maximum of sparse matrix A on dimension `dim`.
```
Comment (jermainewang) [inline python/dgl/mock_sparse/reduce.py:145]:
```suggestion
"""Compute the maximum of non-zero values in sparse matrix A
along the given dimension ``dim``.
```

Comment (jermainewang) [reply python/dgl/mock_sparse/reduce.py:145]: All docstring is in RST. So \`xxx\` is for italic font. Use \`\`xxx\`\` for code.

File: python/dgl/mock_sparse/reduce.py [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+
+    Defining on a SparseMatrix, `smax` does not consider its zero values.
+
+    If `dim` is None, it reduces all the elements in the sparse matrix.
```
Comment (jermainewang) [inline python/dgl/mock_sparse/reduce.py:149]: Use :attr:\`dim\` to mark argument.

File: python/dgl/mock_sparse/reduce.py [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+    >>> print(A.sum(0))
+    tensor([[3, 3], [0, 0], [2, 2]])
+    """
+    return A.reduce(dim, rtype="sum")
+
+
+def smax(A: SparseMatrix, dim=None):
+    """Returns the maximum of sparse matrix A on dimension `dim`.
+
+    Defining on a SparseMatrix, `smax` does not consider its zero values.
```
Comment (jermainewang) [inline python/dgl/mock_sparse/reduce.py:147]: The sentence doesn't parse. Do you mean "The reduction only considers non-zero values."? If so, you have described this below so the sentence can be deleted.

File: python/dgl/mock_sparse/reduce.py [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+    >>> print(A.smax(0))
+    tensor([[2, 2], [0, 0], [2, 2]])
+    >>> print(A.smax(1))
+    tensor([[1, 2], [2, 2], [0, 0], [0, 0]])
+    """
+    return A.reduce(dim, rtype="smax")
+
+
+def smin(A: SparseMatrix, dim=None):
+    """Returns the minimum of sparse matrix A on dimension `dim`.
```
Comment (jermainewang) [inline python/dgl/mock_sparse/reduce.py:196]: Same comment on the docstring. See ``smax``.

File: python/dgl/mock_sparse/reduce.py [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+    >>> print(A.smin(0))
+    tensor([[1, 1], [0, 0], [2, 2]])
+    >>> print(A.smin(1))
+    tensor([[1, 2], [2, 1], [0, 0], [0, 0]])
+    """
+    return A.reduce(dim, rtype="smin")
+
+
+def smean(A: SparseMatrix, dim=None):
+    """Returns the mean of sparse matrix A on dimension `dim`.
```
Comment (jermainewang) [inline python/dgl/mock_sparse/reduce.py:247]: Same comment on the docstring. See ``smax``.

File: python/dgl/mock_sparse/reduce.py (hunk 6 above)
Comment (czkkkkkk) [reply python/dgl/mock_sparse/reduce.py]: Right.

File: python/dgl/mock_sparse/__init__.py (hunk 10 above)
Comment (czkkkkkk) [reply python/dgl/mock_sparse/__init__.py]: Done.

File: python/dgl/mock_sparse/reduce.py (hunk 11 above)
Comment (czkkkkkk) [reply python/dgl/mock_sparse/reduce.py]: Done.

File: python/dgl/mock_sparse/reduce.py (hunk 12 above)
Comment (czkkkkkk) [reply python/dgl/mock_sparse/reduce.py]: Changed.

File: python/dgl/mock_sparse/reduce.py (hunk 13 above)
Comment (czkkkkkk) [reply python/dgl/mock_sparse/reduce.py]: Done.

File: python/dgl/mock_sparse/reduce.py (hunk 14 above)
Comment (czkkkkkk) [reply python/dgl/mock_sparse/reduce.py]: Right. I deleted them.

File: tests/pytorch/mock_sparse/test_reduce.py (hunk 7 above)
Comment (czkkkkkk) [reply tests/pytorch/mock_sparse/test_reduce.py]: Done.

**Review by jermainewang — APPROVED** [review]
LGTM. After you fixed the lint problem, go ahead and merge it.
